Plug-in framework: Cursor methods for seeking

mgavioli opened the pull request:
Implements the following method of the `Cursor` type:
- `scoreStart()`
- `scoreEnd()`
- `selectionStart()`
- `selectionEnd()`

All methods obey the current track and the current filter:
- `xxxStart()` move to first segment of type included in `Cursor.filter` which has an element in the current track.
- `xxxEnd()` move to last segment of type included in `Cursor.filter` which has an element in the current track.

`selectionStart/End()` default to `scoreStart/End()` if there is no selection.
The old `Cursor.rewind(int)` method remains unchanged for legacy compatibility.

Any method may result in `Cursor.segment` being set to `null` if there is no segment satisfying the conditions.

---

Additionally:
- added a test to `tst_scripting` for these new methods
- added comments for the manual to all `Cursor` methods
- added the read-only informative properties `Score.selectionFirstTrack` and `Score.selectionLastTrack`

mgavioli commented:
@Jojo-Schmitz : good catch! Thanks.

lasconic commented:
There are some important changes here. I know @heuchi has been working on several plugins. Maybe you have an opinion?

heuchi commented:
Bug report first ;-)

Cursor.selectionEnd() has several problems:
<ol><li>it sets the cursor to the start of the score if there is no selection.</li>
<li>it sets the cursor to the start of the score if the selection contains the last segment in which case  _segment will also be a null pointer.</li>
<li>it sets the cursor outside the selection, because prevInTrack() will not move the cursor when it already sits on a valid segment type.</li></ol>

However, I tried to recreate the main loop of the standard Cursor-based plugin type we currently use, i.e. iterating over tracks of the selection/whole score. This is what I came up with:

<pre>
   onRun: {
      if (typeof curScore === 'undefined')
            Qt.quit();

      var cursor = curScore.newCursor();

      for (var track = 0; track < curScore.ntracks; track++) {
            cursor.track = track;
            // find end of selection (if present)
            cursor.selectionEnd();
            var endTick = cursor.tick;
            // goto start of selection/score
            cursor.selectionStart();

            while (cursor.segment && cursor.tick <= endTick) {
                  // do work
                  cursor.next();
            }
      }
      Qt.quit();
   }
</pre>

This looks <b>a lot better</b> than what we have now.

The one little thing that does not yet look perfect to me is that you have to call selectionEnd() several times, because if you call it for a track that's not included in the selection, you get 0.

heuchi commented on libmscore/cursor.h:
```diff
@@ -96,18 +97,34 @@ class Cursor : public QObject {
+      //@ adds an Element to the current cursor segment in the current track
+      Q_INVOKABLE void add(Ms::Element*);
+      //@ adds a note of pitch 'pitch' to the chord in current track of current segment;
+      //@ chord must already exist
```
I think the opposite is true: addNote(pitch) will replace whatever is at the current cursor position with a single note with the given pitch.

heuchi commented on libmscore/cursor.h:
```diff
@@ -96,18 +97,34 @@ class Cursor : public QObject {
+      //   type=1      rewind to start of selection
+      //   type=2      rewind to end of selection
+      Q_INVOKABLE void rewind(int type);  // obsolete; replaced by scoreStart/End() and selectionStart/End():
+
```
I probably wouldn't remove the description completely, because rewind() can still be found in existing plugins.
Besides: If you don't intent to use the Cursor for iterating, but only to find the selection, using selectionStart() and selectionEnd() to retrieve the first track, last track, first segment and last segment of the selection doesn't seem to be a lot easier than using rewind().

mgavioli commented:
@heuchi : your bug findings do not match mines. Your cases 1. and 3. are tested by the `mtest/scripting/cursor_seek.qml` test and work as expected. Your case 2. is not tested; I modified the test for it and it works.

@heuchi : `addNote()` : I haven't change any code here, I only updated the plug-in docum. comments to match what the code does (or says to do). From the code for `Score::undoAddElement()`, it seems it does what the comment says. Do you know for sure it does not?

@lasconic : the cursor methods have been reworked according to your suggestions in my old PR https://github.com/musescore/MuseScore/pull/1916

If it can be further improved, please advise. For instance, given the prevalent usage, it may make sense to add methods like `Score.selectionStartTick(track)` and `Score.selectionEndTick(track)` to query the selection boundaries and allow a loop like:

    for (var track = curScore.selectionFirstTrack; track <= curScore.selectionLastTrack; track++)
    {
        cursor.track = track;
        // goto start of selection/score
        cursor.selectionStart();

        while (cursor.segment && cursor.tick <= curScore.selectionEndTick(track))
        {
            // do work
            cursor.next();
        }
    }

Finally, about the old `Cursor.rewind()` method:

1) The idea is not to remove the old `Cursor.rewind()` method, but to leave it undocumented (and implicitly deprecated).

2) There is a basic difference between `Cursor.rewind()` and these new methods: the latter (as far as I can tell) always work as expected, the former not always.

3) The latter are also intended mostly as a clearer and more consistent version of the former, rather than than as something completely different. If using them is not so different, well, this is a bonus, I believe.

4) It does not really makes sense to find the first / last segment of the selection in general, as there may be (and in most cases there are) tracks with nothing in the selection, maybe tracks with nothing at all and the selection boundaries would not make any sense for them.

In addition, as now `Cursor` is segment-type selective, a track may contain something, but not of the type(s) looked for (for instance, any voice != 0 will not contain any bar line or volta or similar). So, going track by track is really necessary.

heuchi commented:
Here's a <a href="https://www.dropbox.com/l/G3tN7N3OdlpxqCpXJz8Qpm">demonstration plugin</a>, that marks the cursor position after calling selectionStart() with an 'S' and the position after calling selectionEnd() with an 'E' for all tracks, if a valid segment is found at these positions in <a href="https://www.dropbox.com/l/43gU0LDTbfMX71OmKRuhWk">this demonstration score</a>.
Unfortunately, the blue rectangles surrounding the selection are not shown in a screen shot, but the notes selected are shown blue.

Issue 1: Nothing selected, so selectionEnd() goes to the beginning of the score:
![demonoselection](https://cloud.githubusercontent.com/assets/2162327/7239246/95fc24fa-e7a8-11e4-816b-c4a6f98b9100.png)

Issue 2: Selection contains last segment, selectionEnd() goes to the beginning of the score:
![demoselectionend](https://cloud.githubusercontent.com/assets/2162327/7239281/f12adc9a-e7a8-11e4-8bdf-efc83ae14d24.png)

Issue 3: selectionEnd() goes to note outside the selection:
![demoselection](https://cloud.githubusercontent.com/assets/2162327/7239285/01f29522-e7a9-11e4-9fee-0c89b3f39789.png)

mgavioli commented:
@heuchi @lasconic  : Yes, there was a problem showing up where there was no selection whatsoever (damn cut&paste!), which the current tests did not check. Fixed and added more tests.

There is also an oddity: if there is a selection and it ends at score end, `Selection::endSegment()` returns `null`; the updated `Cursor` code takes care of it anyway.

heuchi commented:
61f3d20 fixed issues 1 and 2. Issue 3 is still there.

mgavioli commented:
@heuchi that `Selection::endSegment()` is supposed to return the segment **after** the end of the selection was well buried in the comments...

Updated code and tests.

heuchi commented:
Ok, now it looks like it's working as it should. So back to the other topics:

About Cursor.addNote(pitch): 
Yes, I did test it, and it does not work as you described.

About Cursor.rewind():
1) I don't see a reason to not document a function, that exists. And I don't understand the meaning of "implicitly deprecated". Either something is deprecated or it is not.

4) The type of selection we're talking about here, and which your functions deal with, is rectangular. So it does indeed make sense defining it by a first and last track, and a first and last segment.
There are already some plugins out there that need to know these four values and there is no easy way of finding the selection without using the Cursor. Your new functions (including Score.selectionEndTick()) don't seem to offer a nicer way of getting these values, so one will probably still rely on rewind(). One more reason to leave it documented, and why it shouldn't yet be called deprecated. 
And regarding your code example: it is not necessary to have a different value of curScore.selectionEndTick for each track; it won't even make a difference.

lasconic commented:
Do we have a consensus on the behavior of the added methods?

If yes, @mgavioli would you mind squashing the commits?
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html 

Regarding rewind(), the goal is to get rid of it in the next version. I don't really care if it is documented or not but we should mark it as deprecated and discourage the usage in new plugins. The new methods should be enough.

heuchi commented:
I don't know if the added methods always behave as expected, because there are still cases where selectionStart() or selectionEnd() will go to an unselected element (if the user unchecks a voice in the selection filter).

> The new methods should be enough.

It probably depends on the definition of 'enough', but finding out if a score has a selection and if so, the first and last selected segment, seems to be really complicated using only the new methods:
<pre>
  var cursor = curScore.newCursor();
  var hasSelection = false;
  var selectionFirstSegment = null;
  var selectionLastSegment = null;
  for (var track = curScore.selectionFirstTrack; track <= curScore.selectionLastTrack; track++) {
        cursor.track = track;
        cursor.selectionStart();
        if (cursor.segment) {
              if (cursor.element.selected)
                    hasSelection = true;
              if (selectionFirstSegment) {
                    if (cursor.tick < selectionFirstSegment.tick)
                          selectionFirstSegment = cursor.segment;
              } else {
                    selectionFirstSegment = cursor.segment;
              }
        }
        cursor.selectionEnd();
        if (cursor.segment) {
              if (selectionLastSegment) {
                    if (cursor.tick > selectionLastSegment.tick)
                          selectionLastSegment = cursor.segment;
              } else {
                    selectionLastSegment = cursor.segment;
              }
        }
  }
</pre>
Maybe someone finds a better way of doing it, because this looks a lot worse than what rewind() can do. But if there is no better way, I really don't think we should mark a method deprecated, when it is actually so much easier to use...

heuchi commented:
I just noticed that rewind(1) also has some problems in finding the start of the selection.
So selectionStart() indeed works better than rewind(1).

In order to make things as easy as possible and finally be able to put an end to this "never-ending" topic, what about adding the following three properties to the Score object (maybe  in a different PR):
<pre>
bool     hasSelection
Segment* selectionFirstSegment (set to first segment of the score if there is no selection)
Segment* selectionLastSegment (set to last segment of the score if there is no selection)
</pre>
This would provide all the information about the selection one could possibly need in a plugin, the new methods would provide a nice way of iterating track by track, and rewind() could be left (undocumented) just for compatibility.

lasconic commented:
If I understand correctly, we are now on par with rewind(), even if we don't add these 3 properties. That was the whole purpose of this PR in a first place, replacing the not really user friendly rewind() by meaningful and "self explanatory" functions.

What's the use case of getting the first and last segment of a selection? The whole point of having a cursor is to provide a way to go through the score without having to know that.

heuchi commented:
There are tasks that cannot be done with a track-by-track cursor, like:
courtesy accidentals, check for parallels, chordal analyzer ...
Besides, the current cursor does not find annotations if there's not also an element in the same track.
In a word, the cursor is still limited in its functionality.

So, offering a nice way to do a track-by-track iteration is fine, but why limit the whole plugin API to this one special case. Why can't we have both:
A nice way to go through the score without knowing some of the implementation detail on the one hand, but also offering nice ways to access all information in the score, if needed.

heuchi commented:
One more thing, we're not yet having a real nice way of exiting the inner loop when we're outside the selection but still first have to call selectionEnd(), remember the tick value and then selectionStart() before entering the loop over all elements.
@mgavioli suggested adding a selectionEndTick(track) to Score, which won't work, because Score doesn't know the cursor's filter setting. But as I pointed out, that wouldn't be necessary.
I don't insist on retrieving the last segment of the selection. Getting the tick value would also be fine, and that could be done in a property Score.selectionEndTick.
However, I don't see a good alternative to having Score.selectionFirstSegment.

lasconic commented:
Of course we can, but you know why ;) two API means two API to support in the future, to document and to choose from for a plugin developer.

The proposed selectionFirstSegment  & selectionLastSegment could also be on a cursor. no?

heuchi commented:
Yes, of course. I just put it in the same class as selectionFirstTrack and selectionLastTrack which this PR has in Score.

heuchi commented:
Speaking of keeping the API simple and small: Why do we actually need scoreEnd() and selectionEnd()?
Once we have Cursor.selectionEndTick (which, as usual, contains the last tick of the score when there's no selection) all we need to do is this:
<pre>
for (cursor.track = curScore.selectionFirstTrack; cursor.track <= curScore.selectionLastTrack; cursor.track++)
{
    cursor.selectionStart();
    while (cursor.segment && cursor.tick <= cursor.selectionEndTick)
    {
        // do work
        cursor.next();
    }
}
</pre>

mgavioli commented:
@lasconic : squashed and rebased.

@lasconic @heuchi : I have no strong ideas on the topic; please feel free to merge this PR if reputed good or at least a good starting point, to close it if not useful and/or to build upon it.

Jojo-Schmitz commented:
Hmm, if this should (stand a chance of) get cherry picked for 2.0.1, I guess the script testing stuff would need to be in a separate commit?

mgavioli commented:
Rebased to master
